### DISCUSSION TITLE
Embody a deprecation policy for spec services

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#553] [Ideas]
I think the `http-service` removal is a good example for that the current process of remove a spec-service is flawed and actually a removed service can't be "easily recovered anytime" as promised in the latest spec-calls.

I therefore like to propose the following:

If a service is about to removed:

- this should be mentioned in the spec, maybe at the beginning of the spec some kind of "Deprecation notices"
- if there are any service interfaces, there should be a minor version increment and all interfaces, classes, enums, ... are marked with `@Deprecated` plus a reference to an online source that describes the deprecation and how one can get in touch and probably veto against it
- it is then removed in **the next major** specification release if no one tries to address the concerns that lead to the deprecation

That way API consumers and providers have at least a chance to notice whats going on and there is enough time to investigate or get in contact.

**Comment by timothyjward** [comment]
> I think the http-service removal is a good example for that the current process of remove a spec-service is flawed and actually a removed service can't be "easily recovered anytime" as promised in the latest spec-calls.

I don't agree with this statement. The Http Service could be extremely easily resurrected, however that isn't what was being asked for. The expressed desire was for the Http Service to be resurrected *and* updated to the Jakarta namespace.

The decision not to update the Http Service to Jakarta was separate, and based on the fact that nobody had expressed any interest in maintaining or updating the Http Service specification in well over a decade. The work involved in updating the specification would have been substantial, and there was very broad consensus that it should not be attempted.

If there is a strong desire to bring the Http Service back (as it was in R7) in `version.next`  then I wouldn't be opposed to this, but I do still think that there is no value in updating the specification to Jakarta.

**Reply by laeubi** [reply]
> > I think the http-service removal is a good example for that the current process of remove a spec-service is flawed and actually a removed service can't be "easily recovered anytime" as promised in the latest spec-calls.
> 
> I don't agree with this statement. The Http Service could be extremely easily resurrected, however that isn't what was being asked for. The expressed desire was for the Http Service to be resurrected _and_ updated to the Jakarta namespace.

Sorry to say but that's really slitting hairs... so if it has to include two steps (resurrected it then update it), why there are so many discussions and not a plain and simple: That would be possible but we must do it in two steps?

> but I do still think that there is no value in updating the specification to Jakarta

No value for you probably, but there are at least:

- [Equinox](https://github.com/osgi/osgi/issues/551)
- [Camel](https://camel.apache.org/)
- [CXF](https://cxf.apache.org/)
- [Pax Web](https://github.com/ops4j/org.ops4j.pax.web)
- ... probably more ...

currently using `HttpService` and they have their users ...

And in general such decisions have a much bigger impact on the eco-system, if people learn that OSGi services are actually just vanish or are not updated they will loose trust in that OSGi standard service are actually anything one might better not use at all.

**Reply by tjwatson** [reply]
If someone wants to update the original http service specification to use the Jakarta namespace then they should follow the spec process to do that.  I think I missed the call this was discussed, but I see no reason someone interested in doing the work would not be able to drive that.  My current employment does not afford me the time to do that work since it is not needed by the products I work on.  With that said, I would still find time help out where I can for Equinox interests.  But I still find it difficult to justify because it will involve a 2.0 version of the specification.  Absolutely all users of the old http service will be broken by the move to Jakarta.  If the users have to make updates then they may as well migrate to using the Jakarta servlet whiteboard specification.

Instead, what I had thought would be possible is to support the old http servlet (even with its javax.servlet namespace) on top of an jakarta whiteboard implementation.  That involves the use of proxies to get the old http service objects registered with the white board.  So far I've not seen any investigation of doing that (although I think the felix impl may have attempted that).  But I think the Equinox impl could do that since the old http service sits on top of the whiteboard impl already.  Then we can try to migrate users of the old http service to Jakarta whiteboard over time.

**Reply by bjhargrave** [reply]
> Instead, what I had thought would be possible is to support the old http servlet (even with its javax.servlet namespace) on top of an jakarta whiteboard implementation.

This is possible and already done by the Apache Felix Http implementation!

https://github.com/apache/felix-dev/tree/master/http/base/src/main/java/org/apache/felix/http/base/internal/jakartawrappers

**Reply by tjwatson** [reply]
Then I suggest that approach be used to move forward for anyone interested in the old http service currently.  Then make steps to migrate to jakarta whiteboard over time.

**Reply by laeubi** [reply]
Please discuss HttpService concerns here: https://github.com/osgi/osgi/issues/551 :-)
Beside that, an "old" HttpService is quite useless as most code moving to jakarta namespace (**but without any functional difference**!)

This was meant to how make a removal process more visible to the user **before** something is removed!
